<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Комментарий ревьюера</b>


Привет, Сергей! Давай знакомиться! Меня зовут Дмитрий Махортов, и я буду проверять твой проект. Сразу предлагаю общение на «ты» 🙂, но если тебе это не комфортно, то дай знать, и мы перейдем на «вы». 


Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе погрузиться в увлекательный мир работы с данными и вырасти в крепкого профи. Это отдаленная цель. А ближайшая - сделать твою работу еще лучше )).
   
    
Все ключевые этапы в работе выполнены, и я вижу что с проектом ты справшяешься. Есть моменты, которые нужно доработать, но я уверен, у тебя все получится.
        
Расскажу как обычно проходит проверка:    
Бывают моменты, которые требуют пристального внимания. Комментарии по ним выделены <span style='background-color:#F7B3A4'> красным цветом </span> и обозначены значком 🛑. После их доработки проект будет принят.  🙂
 
<span style='background-color:#B7EBA7'> Зеленым цветом </span> и значком ✅ отмечены удачные и элегантные решения, на которые можно опираться в будущих проектах. Или советы «со звездочкой», которые помогут тебе в будущем.

<span style='background-color:#F9EDA6'>Жёлтым цветом </span> и значком ⚠️ выделено то, что в следующий раз можно сделать по-другому. Ты можешь учесть эти комментарии при выполнении будущих заданий или доработать проект сейчас (однако это не обязательно).

Давай работать над проектом в диалоге: **если ты что-то меняешь в проекте по моим рекомендациям — пиши об этом**. Выбери для своих комментариев какой-то заметный цвет, так мне будет легче отследить изменения. Пожалуйста, **не перемещай, не изменяй и не удаляй мои комментарии**. Всё это поможет выполнить повторную проверку твоего проекта оперативнее. 
</div>

<div class="alert" style="background-color:#ead7f7;color:#8737bf">
    <font size="3"><b>образец комментария студента</b></font>
   
Можешь использовать этот комментарий. Просто удали этот текст и вместо него напиши свой 🙂

</div>

# Прекод

# Сборный проект-4

Вам поручено разработать демонстрационную версию поиска изображений по запросу.

Для демонстрационной версии нужно обучить модель, которая получит векторное представление изображения, векторное представление текста, а на выходе выдаст число от 0 до 1 — покажет, насколько текст и картинка подходят друг другу.

### Описание данных

Данные лежат в папке `/datasets/image_search/` или доступны по [ссылке](https://code.s3.yandex.net/datasets/dsplus_integrated_project_4.zip).

В файле `train_dataset.csv` находится информация, необходимая для обучения: имя файла изображения, идентификатор описания и текст описания. Для одной картинки может быть доступно до 5 описаний. Идентификатор описания имеет формат `<имя файла изображения>#<порядковый номер описания>`.

В папке `train_images` содержатся изображения для тренировки модели.

В файле `CrowdAnnotations.tsv` — данные по соответствию изображения и описания, полученные с помощью краудсорсинга. Номера колонок и соответствующий тип данных:

1. Имя файла изображения.
2. Идентификатор описания.
3. Доля людей, подтвердивших, что описание соответствует изображению.
4. Количество человек, подтвердивших, что описание соответствует изображению.
5. Количество человек, подтвердивших, что описание не соответствует изображению.

В файле `ExpertAnnotations.tsv` содержатся данные по соответствию изображения и описания, полученные в результате опроса экспертов. Номера колонок и соответствующий тип данных:

1. Имя файла изображения.
2. Идентификатор описания.

3, 4, 5 — оценки трёх экспертов.

Эксперты ставят оценки по шкале от 1 до 4, где 1 — изображение и запрос совершенно не соответствуют друг другу, 2 — запрос содержит элементы описания изображения, но в целом запрос тексту не соответствует, 3 — запрос и текст соответствуют с точностью до некоторых деталей, 4 — запрос и текст соответствуют полностью.

В файле `test_queries.csv` находится информация, необходимая для тестирования: идентификатор запроса, текст запроса и релевантное изображение. Для одной картинки может быть доступно до 5 описаний. Идентификатор описания имеет формат `<имя файла изображения>#<порядковый номер описания>`.

В папке `test_images` содержатся изображения для тестирования модели.

## 1. Исследовательский анализ данных

Наш датасет содержит экспертные и краудсорсинговые оценки соответствия текста и изображения.

В файле с экспертными мнениями для каждой пары изображение-текст имеются оценки от трёх специалистов. Для решения задачи вы должны эти оценки агрегировать — превратить в одну. Существует несколько способов агрегации оценок, самый простой — голосование большинства: за какую оценку проголосовала большая часть экспертов (в нашем случае 2 или 3), та оценка и ставится как итоговая. Поскольку число экспертов меньше числа классов, может случиться, что каждый эксперт поставит разные оценки, например: 1, 4, 2. В таком случае данную пару изображение-текст можно исключить из датасета.

Вы можете воспользоваться другим методом агрегации оценок или придумать свой.

В файле с краудсорсинговыми оценками информация расположена в таком порядке:

1. Доля исполнителей, подтвердивших, что текст **соответствует** картинке.
2. Количество исполнителей, подтвердивших, что текст **соответствует** картинке.
3. Количество исполнителей, подтвердивших, что текст **не соответствует** картинке.

После анализа экспертных и краудсорсинговых оценок выберите либо одну из них, либо объедините их в одну по какому-то критерию: например, оценка эксперта принимается с коэффициентом 0.6, а крауда — с коэффициентом 0.4.

Ваша модель должна возвращать на выходе вероятность соответствия изображения тексту, поэтому целевая переменная должна иметь значения от 0 до 1.


<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
 👍 Да, это хорошая практика - дать описание контекста и проблемы, которую мы решаем. </div>

In [1]:
#!pip install spacy
#!pip install spacy[transformers]
!pip install -q torchvision
#!python -m spacy download en_core_web_lg
!python -m spacy download en_core_web_sm

#!pip install  spacy download en_core_web_sm
#!pip3 install https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-2.2.0/en_core_web_sm-2.2.0.tar.gz
#!pip install https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-3.0.0/en_core_web_sm-3.0.0.tar.gz
#!pip install https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-3.0.0/en_core_web_sm-3.0.0-py3-none-any.whl

     ---------------------------------------- 0.0/12.8 MB ? eta -:--:--
     --------------------------------------- 0.0/12.8 MB 330.3 kB/s eta 0:00:39
     --------------------------------------- 0.0/12.8 MB 330.3 kB/s eta 0:00:39
     --------------------------------------- 0.0/12.8 MB 245.8 kB/s eta 0:00:52
     --------------------------------------- 0.1/12.8 MB 357.2 kB/s eta 0:00:36
     --------------------------------------- 0.1/12.8 MB 467.6 kB/s eta 0:00:28
      -------------------------------------- 0.2/12.8 MB 618.3 kB/s eta 0:00:21
      -------------------------------------- 0.2/12.8 MB 724.0 kB/s eta 0:00:18
      -------------------------------------- 0.3/12.8 MB 803.7 kB/s eta 0:00:16
     - ------------------------------------- 0.4/12.8 MB 928.4 kB/s eta 0:00:14
     - ------------------------------------- 0.4/12.8 MB 922.1 kB/s eta 0:00:14
     - -------------------------------------- 0.6/12.8 MB 1.2 MB/s eta 0:00:10
     -- ------------------------------------- 0.7

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Лично меня очень утомляют простыни логов после установки библиотек. Я их все равно никто не читает. Можно добавить флаг -q будет гораздо аккуратноее.
    
    !pip install -q torchvision
    
Кстати, я вижу ты делаешь повторную установку torchvision, что для локальной среды уже лишнее. А во многих облачных средах (например в Google Colab) эта библиотека уже установлена.     
</div>

<div class="alert" style="background-color:#ead7f7;color:#8737bf">
    <font size="3"><b>Комментарий студента</b></font>
   
torchvision убрал логи, а  en_core_web_sm не получилось. Может быть еще есть вариант как их убрать?

</div>

In [2]:
import torchvision.models as models
import torch
import torch.nn as nn
import numpy as np
#import tensorflow as tf
import math
import pandas as pd
import random as rd

from PIL import Image
import os

import spacy
import re
import nltk
from nltk.corpus import stopwords
nltk.download('stopwords')
#from nltk.corpus import stopwords as nltk_stopwords
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import GroupShuffleSplit
from sklearn.preprocessing import  StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import make_scorer
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import MinMaxScaler

#from skorch.callbacks import EpochScoring, EarlyStopping
#from skorch import NeuralNetRegressor
from sklearn.pipeline import Pipeline
from tqdm import tqdm, tqdm_notebook
#import torch
#import torch.nn as nn
from math import ceil

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\Serge\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Отлично, все нужные библиотеки импортированы в начале ноутбука.Это хорошая практика.</div>

Импортируем данные.

In [3]:
train_dataset = pd.read_csv('to_upload/train_dataset.csv')
CrowdAnnotations = pd.read_csv('to_upload/CrowdAnnotations.tsv', sep='\t', header=None)
ExpertAnnotations = pd.read_csv('to_upload/ExpertAnnotations.tsv', sep='\t', header=None)
test_queries = pd.read_csv('to_upload/test_queries.csv', sep='|', index_col=0)
test_images = pd.read_csv('to_upload/test_images.csv', sep='|')

<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Комментарий ревьюера</b>

Данный проект очень хороший пример того, как один и тот же код может запускаться на разных платформах, у разных людей, даже на операционных системах. Поэтому лучше заранее позаботиться о том, чтобы при переходе с одного компьютера на другой нужно было вносить как можно меньше изменений. Самый простой способ - вынести путь к данным в константу, которую определлить в самом начале ноутбука и все пути определять с этой константой.
    
Также для объединения путей лучше испльзовать os.path.join, т.к. этот метод соединяет пути с учетом операционной системы (в linux и windows для путей используются знаки / и \)
    
Если ты используешь colab или другой облачный сервис, то датасет в виртуальную машину можно загрузить вот так
    
    
    # загрузка датасета 

    !wget -q https://code.s3.yandex.net/datasets/dsplus_integrated_project_4.zip
    !unzip -q dsplus_integrated_project_4.zip

А вот это образец универсального кода для загрузки данных    
    
    DATA_PATH = твой путь    
    
    df_train = pd.read_csv(os.path.join(DATA_PATH, 'train_dataset.csv'))
    ....

 
    
    


</div>

In [4]:
train_dataset.sample(10)

image                     query_id  \
726   1220401002_3f44b1f3f7.jpg  3470951932_27ed74eb0b.jpg#2   
637   3388330419_85d72f7cda.jpg  2228022180_9597b2a458.jpg#2   
2269  1432179046_8e3d75cf81.jpg  3708177171_529bb4ff1d.jpg#2   
4386  2484190118_e89363c465.jpg   270816949_ffad112278.jpg#2   
2277  3027397797_4f1d305ced.jpg  1174525839_7c1e6cfa86.jpg#2   
1160  1287475186_2dee85f1a5.jpg  2599444370_9e40103027.jpg#2   
3062  3110649716_c17e14670e.jpg  3244747165_17028936e0.jpg#2   
1598  2654514044_a70a6e2c21.jpg  2862004252_53894bb28b.jpg#2   
3870  2525270674_4ab536e7ec.jpg  2890113532_ab2003d74e.jpg#2   
5649  3353036763_4cbeba03b2.jpg  1897025969_0c41688fa6.jpg#2   

                                             query_text  
726                     A boy runs to a baseball base .  
637   Group of children in blue uniforms sitting on ...  
2269  A skateboarder is performing a jump from a woo...  
4386  A person riding a horse with a mountain range ...  
2277  A young , long-haired , girl on the beach , is...  
1160  Two black dogs are dog paddling in a sparkling...  
3062  The young adults take a group shot of themselv...  
1598  A brown and tan dog walks through the green gr...  
3870  Two dogs in motion in a grassy field with smal...  
5649  A black dog bites a white dog while standing o...

In [5]:
train_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5822 entries, 0 to 5821
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   image       5822 non-null   object
 1   query_id    5822 non-null   object
 2   query_text  5822 non-null   object
dtypes: object(3)
memory usage: 136.6+ KB


In [6]:
train_dataset.duplicated().sum()

0

В тренировочном датасете 5822 строк, или в среднем 5,8 описаний  на каждую из 1000 изображений.

In [7]:
train_dataset.describe()

image                     query_id  \
count                        5822                         5822   
unique                       1000                          977   
top     3364151356_eecd07a23e.jpg  2600867924_cd502fc911.jpg#2   
freq                           10                           34   

                                               query_text  
count                                                5822  
unique                                                977  
top     Two dogs , one brown and white and one black a...  
freq                                                   34

При этом уникальных запросов  977

In [8]:
CrowdAnnotations = pd.read_csv('to_upload/ExpertAnnotations.tsv',  sep='\t', header=None, names=['image', 'query_id', 'percentage_confirmed', 'quantity_confirmed', 'quantity_disproved'])
CrowdAnnotations.head(10)

image                     query_id  \
0  1056338697_4f7d7ce270.jpg  2549968784_39bfbe44f9.jpg#2   
1  1056338697_4f7d7ce270.jpg  2718495608_d8533e3ac5.jpg#2   
2  1056338697_4f7d7ce270.jpg  3181701312_70a379ab6e.jpg#2   
3  1056338697_4f7d7ce270.jpg  3207358897_bfa61fa3c6.jpg#2   
4  1056338697_4f7d7ce270.jpg  3286822339_5535af6b93.jpg#2   
5  1056338697_4f7d7ce270.jpg  3360930596_1e75164ce6.jpg#2   
6  1056338697_4f7d7ce270.jpg  3545652636_0746537307.jpg#2   
7  1056338697_4f7d7ce270.jpg   434792818_56375e203f.jpg#2   
8   106490881_5a2dd9b7bd.jpg  1425069308_488e5fcf9d.jpg#2   
9   106490881_5a2dd9b7bd.jpg  1714316707_8bbaa2a2ba.jpg#2   

   percentage_confirmed  quantity_confirmed  quantity_disproved  
0                     1                   1                   1  
1                     1                   1                   2  
2                     1                   1                   2  
3                     1                   2                   2  
4                     1                   1                   2  
5                     1                   1                   1  
6                     1                   1                   1  
7                     1                   1                   2  
8                     1                   1                   1  
9                     2                   2                   2

In [9]:
CrowdAnnotations.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5822 entries, 0 to 5821
Data columns (total 5 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   image                 5822 non-null   object
 1   query_id              5822 non-null   object
 2   percentage_confirmed  5822 non-null   int64 
 3   quantity_confirmed    5822 non-null   int64 
 4   quantity_disproved    5822 non-null   int64 
dtypes: int64(3), object(2)
memory usage: 227.6+ KB


In [10]:
CrowdAnnotations.duplicated().sum()

0

In [11]:
CrowdAnnotations.describe()

percentage_confirmed  quantity_confirmed  quantity_disproved
count           5822.000000         5822.000000         5822.000000
mean               1.436620            1.624356            1.881999
std                0.787084            0.856222            0.904087
min                1.000000            1.000000            1.000000
25%                1.000000            1.000000            1.000000
50%                1.000000            1.000000            2.000000
75%                2.000000            2.000000            2.000000
max                4.000000            4.000000            4.000000

Около 9% из опрошенных подтверждают, что описание соответствует изображению. При этом, имеется максимальное значение — 1, что означает, что есть изображения с описанием, релевантность которого подвердили абсолютно все "краудсорсеры".

In [12]:
ExpertAnnotations = pd.read_csv('to_upload/ExpertAnnotations.tsv', sep='\t', header=None, names=['image', 'query_id', 'expert_1', 'expert_2', 'expert_3'])
ExpertAnnotations.head(10)

image                     query_id  expert_1  expert_2  \
0  1056338697_4f7d7ce270.jpg  2549968784_39bfbe44f9.jpg#2         1         1   
1  1056338697_4f7d7ce270.jpg  2718495608_d8533e3ac5.jpg#2         1         1   
2  1056338697_4f7d7ce270.jpg  3181701312_70a379ab6e.jpg#2         1         1   
3  1056338697_4f7d7ce270.jpg  3207358897_bfa61fa3c6.jpg#2         1         2   
4  1056338697_4f7d7ce270.jpg  3286822339_5535af6b93.jpg#2         1         1   
5  1056338697_4f7d7ce270.jpg  3360930596_1e75164ce6.jpg#2         1         1   
6  1056338697_4f7d7ce270.jpg  3545652636_0746537307.jpg#2         1         1   
7  1056338697_4f7d7ce270.jpg   434792818_56375e203f.jpg#2         1         1   
8   106490881_5a2dd9b7bd.jpg  1425069308_488e5fcf9d.jpg#2         1         1   
9   106490881_5a2dd9b7bd.jpg  1714316707_8bbaa2a2ba.jpg#2         2         2   

   expert_3  
0         1  
1         2  
2         2  
3         2  
4         2  
5         1  
6         1  
7         2  
8         1  
9         2

In [13]:
ExpertAnnotations.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5822 entries, 0 to 5821
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   image     5822 non-null   object
 1   query_id  5822 non-null   object
 2   expert_1  5822 non-null   int64 
 3   expert_2  5822 non-null   int64 
 4   expert_3  5822 non-null   int64 
dtypes: int64(3), object(2)
memory usage: 227.6+ KB


In [14]:
ExpertAnnotations.duplicated().sum()

0

In [15]:
ExpertAnnotations.describe()

expert_1     expert_2     expert_3
count  5822.000000  5822.000000  5822.000000
mean      1.436620     1.624356     1.881999
std       0.787084     0.856222     0.904087
min       1.000000     1.000000     1.000000
25%       1.000000     1.000000     1.000000
50%       1.000000     1.000000     2.000000
75%       2.000000     2.000000     2.000000
max       4.000000     4.000000     4.000000

Третий эксперт оценивает сходство запроса и картинки выше остальных экспертов (среднее 1,88 против 1,4 и 1,6) 

In [16]:
test_queries = pd.read_csv('to_upload/test_queries.csv', sep='|', index_col=0)
test_queries.head(10)

query_id  \
0  1177994172_10d143cb8d.jpg#0   
1  1177994172_10d143cb8d.jpg#1   
2  1177994172_10d143cb8d.jpg#2   
3  1177994172_10d143cb8d.jpg#3   
4  1177994172_10d143cb8d.jpg#4   
5  1232148178_4f45cc3284.jpg#0   
6  1232148178_4f45cc3284.jpg#1   
7  1232148178_4f45cc3284.jpg#2   
8  1232148178_4f45cc3284.jpg#3   
9  1232148178_4f45cc3284.jpg#4   

                                          query_text  \
0  Two blonde boys , one in a camouflage shirt an...   
1  Two boys are squirting water guns at each other .   
2            Two boys spraying each other with water   
3  Two children wearing jeans squirt water at eac...   
4  Two young boys are squirting water at each oth...   
5                    A baby girl playing at a park .   
6  A closeup of a child on a playground with adul...   
7  A young boy poses for a picture in front of a ...   
8  A young girl is smiling in front of the camera...   
9  There is a little blond hair girl with a green...   

                       image  
0  1177994172_10d143cb8d.jpg  
1  1177994172_10d143cb8d.jpg  
2  1177994172_10d143cb8d.jpg  
3  1177994172_10d143cb8d.jpg  
4  1177994172_10d143cb8d.jpg  
5  1232148178_4f45cc3284.jpg  
6  1232148178_4f45cc3284.jpg  
7  1232148178_4f45cc3284.jpg  
8  1232148178_4f45cc3284.jpg  
9  1232148178_4f45cc3284.jpg

In [17]:
test_queries.info()

<class 'pandas.core.frame.DataFrame'>
Index: 500 entries, 0 to 499
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   query_id    500 non-null    object
 1   query_text  500 non-null    object
 2   image       500 non-null    object
dtypes: object(3)
memory usage: 15.6+ KB


In [18]:
test_queries.duplicated().sum()

0

In [19]:
test_queries.describe()

query_id  \
count                           500   
unique                          500   
top     1177994172_10d143cb8d.jpg#0   
freq                              1   

                                               query_text  \
count                                                 500   
unique                                                500   
top     Two blonde boys , one in a camouflage shirt an...   
freq                                                    1   

                            image  
count                         500  
unique                        100  
top     1177994172_10d143cb8d.jpg  
freq                            5

Видим 100 изображений, по 5 запросов на каждое.

### ВЫВОДЫ

 * Проведена загрузка данных и созданы переменные для них.
 * Обнаружено, что в работе на краудсорсинге соответствие текстового запроса изображению отмечается менее, чем в 9% случаев.
 * Эксперты в целом также невысоко оценивают соответствие текста изображению, их средние оценки колеблются в диапазоне 1,4-1,9 баллов из 4 и можно сделать вывод что запрос и изображение непохожи.
 * Экспертных оценок у нас 5822 запросов × 3 эксперта, а краудсорсинговых 47830.

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
Ты познакомился с данными. Пожалуйста, напиши выводы что ты в них увидел.

</div>


<div class="alert" style="background-color:#ead7f7;color:#8737bf">
    <font size="3"><b>Комментарий студента</b></font>
   
Дополнено выводами

</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
    
Прежде чем приступать к чистке советую тебе провести предобработку. Общая последовательность шагов следующая:
    
 - вначале мы знакомимся с данными и мысленно составляем план, как мы будем "жонглировать" доступными данными, чтобы получить итоговый датафрейм, на котором будем учить модели
 - в рамках реализации этого плана получаем единый датафрейм используя доступные три.
    
В следующем комментарии подсвечу тебе возможные стратегии использования оценок с плюсами и минусами    
    
    
</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
    
В этом проекте есть несколько возможных стратегий работы с обучающими и краудсорс оценками,  вот эти стратегии с плюсами и минусами:
   
    
 - использовать соединение через outer, в этом случае, при корректной обработке пропусков мы можем получить более 50К размеченых пар. К плюсам данного подхода можно отнести большое количество данных. Минус - в основном это будут данные, размеченные краудсорсерами, а там качество раметки ниже.
 - использовать только экспертные оценки. плюс - высокое качество данных (разметка имеет шкалу). Минус - данных меньше.  Поясню по поводу качества разметки на примере: если на изображении одна собака, а в описании "две собаки бегут по берегу", то эксперты поставять соответствие 0,6-0,7 (переводя в шкалу 0-1), а краудсорсеры поставят 0.  Но на экспертных оценках модель хотя-бы научится находить собак, а на краудсор оценках ничему не начится.
    
 - использование соединения через left.  Данный подход очевидно проигрывает второму варианту: данных столько же, нужно дополнительно возиться с объединением, но качество разметки снижается.. Но это соответсвует предложениям авторов проекта, поэтому такой подход имеет право на жизнь
    
 - объединение через inner. Тут все плохо - и данных ОЧЕНЬ мало (в два раза меньше, чем использовать только экспертные оценки) и качество у них "подпорчено". И навыков по обработке пропусков не получаем.....
 
</div>

## 2. Проверка данных

В некоторых странах, где работает ваша компания, действуют ограничения по обработке изображений: поисковым сервисам и сервисам, предоставляющим возможность поиска, запрещено без разрешения родителей или законных представителей предоставлять любую информацию, в том числе, но не исключительно тексты, изображения, видео и аудио, содержащие описание, изображение или запись голоса детей. Ребёнком считается любой человек, не достигший 16 лет.

В вашем сервисе строго следуют законам стран, в которых работают. Поэтому при попытке посмотреть изображения, запрещённые законодательством, вместо картинок показывается дисклеймер:

> This image is unavailable in your country in compliance with local laws
>

Однако у вас в PoC нет возможности воспользоваться данным функционалом. Поэтому все изображения, которые нарушают данный закон, нужно удалить из обучающей выборки.

In [20]:
# Удаляем пары  с изображениями  и запросами, где от всех трёх экспертов получены разные оценки
unique_votes = ExpertAnnotations.loc[:,['expert_1', 'expert_2', 'expert_3']].apply(pd.Series.nunique, axis=1)
mask = unique_votes <= 2
ExpertAnnotations = ExpertAnnotations[mask]

In [21]:
#Подсчитываем количество голосов каждого эксперта за каждую из оценок
votes = ExpertAnnotations.apply(pd.Series.value_counts, axis=1)

# Находим оценки, за которые проголосовала бОльшая часть экспертов
majority = votes.idxmax(axis=1)

# Содаем новый столбец для этих оценок
ExpertAnnotations['majority'] = majority
ExpertAnnotations

image                     query_id  expert_1  \
0     1056338697_4f7d7ce270.jpg  2549968784_39bfbe44f9.jpg#2         1   
1     1056338697_4f7d7ce270.jpg  2718495608_d8533e3ac5.jpg#2         1   
2     1056338697_4f7d7ce270.jpg  3181701312_70a379ab6e.jpg#2         1   
3     1056338697_4f7d7ce270.jpg  3207358897_bfa61fa3c6.jpg#2         1   
4     1056338697_4f7d7ce270.jpg  3286822339_5535af6b93.jpg#2         1   
...                         ...                          ...       ...   
5817   997722733_0cb5439472.jpg  2981702521_2459f2c1c4.jpg#2         1   
5818   997722733_0cb5439472.jpg  2985679744_75a7102aab.jpg#2         1   
5819   997722733_0cb5439472.jpg  3150742439_b8a352e1e0.jpg#2         1   
5820   997722733_0cb5439472.jpg  3375070563_3c290a7991.jpg#2         1   
5821   997722733_0cb5439472.jpg   486917990_72bd4069af.jpg#2         3   

      expert_2  expert_3  majority  
0            1         1         1  
1            1         2         1  
2            1         2         1  
3            2         2         2  
4            1         2         1  
...        ...       ...       ...  
5817         1         1         1  
5818         1         1         1  
5819         1         2         1  
5820         1         1         1  
5821         3         3         3  

[5696 rows x 6 columns]

Далее  таблицы с оценками экспертов и краудсорсеров по столбцам 'query_id' и 'image'. 

После этого пишем функцию в которой оценка эксперта будет приниматься с коэффициентом 0.7, а оценка крауда — с коэффициентом 0.3. При отсутствии оценки от крауда экспертная оценка принимается с коэффициентом 1.

In [22]:
# Объединим таблицы
df = pd.merge(ExpertAnnotations, CrowdAnnotations, on=['image', 'query_id'], how='left').drop(['expert_1', 'expert_2', 'expert_3'], axis=1)

In [23]:
# функция для агрегирования оценок экспертов и крауда
def avg_note(row):
    majority = row.loc['majority']
    percentage_confirmed = row.loc['percentage_confirmed']
    #проверяем наличие оценки от крауда в каждой строоке
    if math.isnan(percentage_confirmed) :
        note = (majority-1)/3
    else:
        note = (majority-1)/3*0.7 + percentage_confirmed*0.3
    return note

df['avg_note']=df.apply(avg_note, axis=1)
df = df.drop(['majority', 'percentage_confirmed', 'quantity_confirmed', 'quantity_disproved'], axis=1)
df

image                     query_id  avg_note
0     1056338697_4f7d7ce270.jpg  2549968784_39bfbe44f9.jpg#2  0.300000
1     1056338697_4f7d7ce270.jpg  2718495608_d8533e3ac5.jpg#2  0.300000
2     1056338697_4f7d7ce270.jpg  3181701312_70a379ab6e.jpg#2  0.300000
3     1056338697_4f7d7ce270.jpg  3207358897_bfa61fa3c6.jpg#2  0.533333
4     1056338697_4f7d7ce270.jpg  3286822339_5535af6b93.jpg#2  0.300000
...                         ...                          ...       ...
5691   997722733_0cb5439472.jpg  2981702521_2459f2c1c4.jpg#2  0.300000
5692   997722733_0cb5439472.jpg  2985679744_75a7102aab.jpg#2  0.300000
5693   997722733_0cb5439472.jpg  3150742439_b8a352e1e0.jpg#2  0.300000
5694   997722733_0cb5439472.jpg  3375070563_3c290a7991.jpg#2  0.300000
5695   997722733_0cb5439472.jpg   486917990_72bd4069af.jpg#2  1.366667

[5696 rows x 3 columns]

In [24]:
df.describe()

avg_note
count  5696.000000
mean      0.575193
std       0.426619
min       0.300000
25%       0.300000
50%       0.300000
75%       0.833333
max       1.900000

In [25]:
train_dataset['query_text'] = train_dataset['query_text'].str.lower()

In [26]:
disabled_pipes = ["parser",  "ner"]
nlp = spacy.load("en_core_web_sm", disable=disabled_pipes)

In [27]:
def clear_text(text):
    processed_text = re.sub(r'[^A-Za-z ]', ' ', text)
    return ' '.join(processed_text.split())

In [28]:
tqdm.pandas()
train_dataset['query_text'] = train_dataset['query_text'].progress_apply(lambda x: clear_text(x))

100%|██████████████████████████████████████████████████████████████████████████| 5822/5822 [00:00<00:00, 278319.97it/s]


In [29]:
train_dataset['query_text'] = [ " ".join([i.lemma_ for i in doc]) for doc in tqdm(nlp.pipe(train_dataset['query_text'].values, disable=['ner', 'parser']), total=train_dataset.shape[0])]

100%|████████████████████████████████████████████████████████████████████████████| 5822/5822 [00:04<00:00, 1251.01it/s]


In [30]:
def lemmatize(texts):
    lemmatized_texts = []
    for doc in tqdm(nlp.pipe(texts, disable=['ner', 'parser']), total=len(texts)):
        lemmatized_texts.append(" ".join([i.lemma_ for i in doc]))
    return lemmatized_texts

In [31]:
stop_kids = ['child', 'children', 'baby', 'babies', 'girl', 'boy', 'girls', 'boys', 'kid', 'kids', 'juvenile',
             'minor', 'minors', 'teen', 'teens', 'teenager', 'teenagers', 'underage', 'infant', 'infants', 'toddler', 'toddlers']

In [32]:
def check_kids(row):
    text = row.loc['query_text']
    for word in text.split():
        if word in stop_kids:
            return True
    return False

In [33]:
# (добавляем столбец) создаем список/таблицу с признаком наличия "запрещенных" слов в тексте
train_dataset['kids_list'] = train_dataset.apply(check_kids, axis=1)

In [34]:
train_dataset.head(30)

image                     query_id  \
0   1056338697_4f7d7ce270.jpg  2549968784_39bfbe44f9.jpg#2   
1   1262583859_653f1469a9.jpg  2549968784_39bfbe44f9.jpg#2   
2   2447284966_d6bbdb4b6e.jpg  2549968784_39bfbe44f9.jpg#2   
3   2549968784_39bfbe44f9.jpg  2549968784_39bfbe44f9.jpg#2   
4   2621415349_ef1a7e73be.jpg  2549968784_39bfbe44f9.jpg#2   
5   3030566410_393c36a6c5.jpg  2549968784_39bfbe44f9.jpg#2   
6   3155451946_c0862c70cb.jpg  2549968784_39bfbe44f9.jpg#2   
7   3222041930_f642f49d28.jpg  2549968784_39bfbe44f9.jpg#2   
8    343218198_1ca90e0734.jpg  2549968784_39bfbe44f9.jpg#2   
9   3718964174_cb2dc1615e.jpg  2549968784_39bfbe44f9.jpg#2   
10  3741827382_71e93298d0.jpg  2549968784_39bfbe44f9.jpg#2   
11   486712504_36be449055.jpg  2549968784_39bfbe44f9.jpg#2   
12  1056338697_4f7d7ce270.jpg  2718495608_d8533e3ac5.jpg#2   
13  1167669558_87a8a467d6.jpg  2718495608_d8533e3ac5.jpg#2   
14  2616643090_4f2d2d1a44.jpg  2718495608_d8533e3ac5.jpg#2   
15  2648165716_02e2e74fd6.jpg  2718495608_d8533e3ac5.jpg#2   
16  2718495608_d8533e3ac5.jpg  2718495608_d8533e3ac5.jpg#2   
17  3244747165_17028936e0.jpg  2718495608_d8533e3ac5.jpg#2   
18  3503689049_63212220be.jpg  2718495608_d8533e3ac5.jpg#2   
19  3506560025_8d0f4f9ac4.jpg  2718495608_d8533e3ac5.jpg#2   
20  3718964174_cb2dc1615e.jpg  2718495608_d8533e3ac5.jpg#2   
21   700884207_d3ec546494.jpg  2718495608_d8533e3ac5.jpg#2   
22  1056338697_4f7d7ce270.jpg  3181701312_70a379ab6e.jpg#2   
23  3187395715_f2940c2b72.jpg  3181701312_70a379ab6e.jpg#2   
24   463978865_c87c6ca84c.jpg  3181701312_70a379ab6e.jpg#2   
25   488590040_35a3e96c89.jpg  3181701312_70a379ab6e.jpg#2   
26   534875358_6ea30d3091.jpg  3181701312_70a379ab6e.jpg#2   
27  1056338697_4f7d7ce270.jpg  3207358897_bfa61fa3c6.jpg#2   
28  3208074567_ac44aeb3f3.jpg  3207358897_bfa61fa3c6.jpg#2   
29  3251976937_20625dc2b8.jpg  3207358897_bfa61fa3c6.jpg#2   

                                           query_text  kids_list  
0   a young child be wear blue goggle and sit in a...       True  
1   a young child be wear blue goggle and sit in a...       True  
2   a young child be wear blue goggle and sit in a...       True  
3   a young child be wear blue goggle and sit in a...       True  
4   a young child be wear blue goggle and sit in a...       True  
5   a young child be wear blue goggle and sit in a...       True  
6   a young child be wear blue goggle and sit in a...       True  
7   a young child be wear blue goggle and sit in a...       True  
8   a young child be wear blue goggle and sit in a...       True  
9   a young child be wear blue goggle and sit in a...       True  
10  a young child be wear blue goggle and sit in a...       True  
11  a young child be wear blue goggle and sit in a...       True  
12      a girl wear a yellow shirt and sunglass smile       True  
13      a girl wear a yellow shirt and sunglass smile       True  
14      a girl wear a yellow shirt and sunglass smile       True  
15      a girl wear a yellow shirt and sunglass smile       True  
16      a girl wear a yellow shirt and sunglass smile       True  
17      a girl wear a yellow shirt and sunglass smile       True  
18      a girl wear a yellow shirt and sunglass smile       True  
19      a girl wear a yellow shirt and sunglass smile       True  
20      a girl wear a yellow shirt and sunglass smile       True  
21      a girl wear a yellow shirt and sunglass smile       True  
22       a man sleep under a blanket on a city street      False  
23       a man sleep under a blanket on a city street      False  
24       a man sleep under a blanket on a city street      False  
25       a man sleep under a blanket on a city street      False  
26       a man sleep under a blanket on a city street      False  
27  a woman play with long red ribbon in an empty ...      False  
28  a woman play with long red ribbon in an empty ...      False  
29  a woman play with long red ribbon in an empty ...      False

In [35]:
# Фильтрация строк, где kids_list = True
kids_df = train_dataset[train_dataset['kids_list'] == True]
# Составление списка query_id
query_id_kids = kids_df['query_id']

In [36]:
query_id_kids

0       2549968784_39bfbe44f9.jpg#2
1       2549968784_39bfbe44f9.jpg#2
2       2549968784_39bfbe44f9.jpg#2
3       2549968784_39bfbe44f9.jpg#2
4       2549968784_39bfbe44f9.jpg#2
                   ...             
5802     523985664_c866af4850.jpg#2
5803    1352410176_af6b139734.jpg#2
5804    1352410176_af6b139734.jpg#2
5814    2475162978_2c51048dca.jpg#2
5817    2061144717_5b3a1864f0.jpg#2
Name: query_id, Length: 1559, dtype: object

Удаляем последние два символа для получения списка названий файлов.

In [37]:
query_id_kidslist = [query_id[:-2] for query_id in query_id_kids]

In [38]:
len(query_id_kidslist)

1559

Удаляем дубликаты названий

In [39]:
modified_query_id_list = list(set(query_id_kidslist))

In [ ]:
len(modified_query_id_list)

In [40]:
train_dataset_nokids = train_dataset[~train_dataset['image'].isin(modified_query_id_list)]

In [41]:
train_dataset_nokids = train_dataset_nokids.reset_index(drop=True)

In [42]:
train_dataset_nokids

image                     query_id  \
0     1056338697_4f7d7ce270.jpg  2549968784_39bfbe44f9.jpg#2   
1     1262583859_653f1469a9.jpg  2549968784_39bfbe44f9.jpg#2   
2     2621415349_ef1a7e73be.jpg  2549968784_39bfbe44f9.jpg#2   
3     3155451946_c0862c70cb.jpg  2549968784_39bfbe44f9.jpg#2   
4     3222041930_f642f49d28.jpg  2549968784_39bfbe44f9.jpg#2   
...                         ...                          ...   
4083   561417861_8e25d0c0e8.jpg   410453140_5401bf659a.jpg#2   
4084   561940436_64d6fc125d.jpg  3157847991_463e006a28.jpg#2   
4085    70995350_75d0698839.jpg  3200120942_59cfbb3437.jpg#2   
4086   909808296_23c427022d.jpg  2112921744_92bf706805.jpg#2   
4087   929679367_ff8c7df2ee.jpg  3651971126_309e6a5e22.jpg#2   

                                             query_text  kids_list  
0     a young child be wear blue goggle and sit in a...       True  
1     a young child be wear blue goggle and sit in a...       True  
2     a young child be wear blue goggle and sit in a...       True  
3     a young child be wear blue goggle and sit in a...       True  
4     a young child be wear blue goggle and sit in a...       True  
...                                                 ...        ...  
4083  a man stand at the entrance of yerba buena cen...      False  
4084    two man in black jacket behind a white handrail      False  
4085  one person be on a ski lift while another snow...      False  
4086          a dog stand on the side of a grassy cliff      False  
4087                          a blurry photo of two dog      False  

[4088 rows x 4 columns]

Таким образом, мы выполнили фильтрацию тренировочного датасета от изображений с детьми. Осталось 4088 строк из 5818, удалено 1730 строк.

Поскольку финальная модель должна обучаться как на текстовых описаниях и оценках, так и на самих изображениях, соединим df, где есть посчитанная по формуле оценка, и train_dataset_nokids, где есть текст запроса, и при этом нет изображений с детьми. Джойн будем производить по паре значений 'image' и 'query_id'.

In [43]:
train_df = pd.merge(df, train_dataset_nokids, on=['image', 'query_id'], how='inner')
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4002 entries, 0 to 4001
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   image       4002 non-null   object 
 1   query_id    4002 non-null   object 
 2   avg_note    4002 non-null   float64
 3   query_text  4002 non-null   object 
 4   kids_list   4002 non-null   bool   
dtypes: bool(1), float64(1), object(3)
memory usage: 129.1+ KB


### !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
Ты использовал соединение через inner и как следствие потерял слишком много данных... Пожалуйста выбери любую другую страгетию (в комментарии выше я подробно расписал тебе разные варианты..</div>


<div class="alert" style="background-color:#ead7f7;color:#8737bf">
    <font size="3"><b>Комментарий студента</b></font>
   
Исправлено

</div>

## 3. Векторизация изображений

Перейдём к векторизации изображений.

Самый примитивный способ — прочесть изображение и превратить полученную матрицу в вектор. Такой способ нам не подходит: длина векторов может быть сильно разной, так как размеры изображений разные. Поэтому стоит обратиться к свёрточным сетям: они позволяют "выделить" главные компоненты изображений. Как это сделать? Нужно выбрать какую-либо архитектуру, например ResNet-18, посмотреть на слои и исключить полносвязные слои, которые отвечают за конечное предсказание. При этом можно загрузить модель данной архитектуры, предварительно натренированную на датасете ImageNet.

Для векторизации возьмем архитектуру ResNet-18, исключив полносвязные слои, которые отвечают за конечное предсказание:

In [44]:
resnet = models.resnet18(pretrained=True)

for param in resnet.parameters():
    param.requires_grad_(False)

print(list(resnet.children()))

modules = list(resnet.children())[:-1]
resnet = nn.Sequential(*modules)

resnet.eval()

C:\Users\Serge\anaconda3\Lib\site-packages\torchvision\models\_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
C:\Users\Serge\anaconda3\Lib\site-packages\torchvision\models\_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=ResNet18_Weights.IMAGENET1K_V1`. You can also use `weights=ResNet18_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)


[Conv2d(3, 64, kernel_size=(7, 7), stride=(2, 2), padding=(3, 3), bias=False), BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True), ReLU(inplace=True), MaxPool2d(kernel_size=3, stride=2, padding=1, dilation=1, ceil_mode=False), Sequential(
  (0): BasicBlock(
    (conv1): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
    (bn1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (relu): ReLU(inplace=True)
    (conv2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
    (bn2): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  )
  (1): BasicBlock(
    (conv1): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
    (bn1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (relu): ReLU(inplace=True)
    (conv2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bia

Sequential(
  (0): Conv2d(3, 64, kernel_size=(7, 7), stride=(2, 2), padding=(3, 3), bias=False)
  (1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (2): ReLU(inplace=True)
  (3): MaxPool2d(kernel_size=3, stride=2, padding=1, dilation=1, ceil_mode=False)
  (4): Sequential(
    (0): BasicBlock(
      (conv1): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU(inplace=True)
      (conv2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn2): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    )
    (1): BasicBlock(
      (conv1): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU(inplace=True)
      (conv2): Con

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Ух, какая "портянка". Ты точно хочешь выводить все слои модели на несколько экранов?</div>

In [45]:
from torchvision import transforms
norm = transforms.Normalize(
    mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])

preprocess = transforms.Compose([
    transforms.Resize(150),
    transforms.CenterCrop(140),
    transforms.ToTensor(),
    norm,
])

In [46]:
image_vectors = []

def image_vect(name, path):
    image = Image.open(path+name).convert('RGB')
    image_tensor = preprocess(image)
    image_tensor = image_tensor.unsqueeze(0)
    output_tensor = resnet(image_tensor).flatten()
    image_vectors.append(output_tensor)

In [47]:
train_df['image'].progress_apply(lambda x: image_vect(x, "to_upload/train_images/"))
outputs_tensor = torch.stack(image_vectors)

100%|██████████████████████████████████████████████████████████████████████████████| 4002/4002 [02:34<00:00, 25.84it/s]


In [48]:
outputs_tensor.shape

torch.Size([4002, 512])

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
 Отлично, вектра для изображений получены👍 </div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Есть технический момент, связанный с оптимальнойстью кода. Конечно, сейчас у тебя слишком маленький датасет и это возможно не критично, но если ты будешь работать с датасетом крупнее, это может стать проблемой. Вообще у нас должно получиться меньше 1000 уникальных изображений (убираем детские). Но в датасете может быть до 40 тысяч записей (в зависимости от выбранной сттарегии), а значит каждое изображение векторизуется по нескольку раз. Чтобы победить это  нужно сильно переписывать код. Но можно использовать декоратор @lru_cahe (попробуй загуглить про него сама, информации очень много). 
    
Ну и конечно же включение GPU тоже может дать вклад в ускорение векторизации.
    

    from functools import lru_cache
    from tqdm.notebook import tqdm
    tqdm.pandas()    

    @lru_cache(1000)
    def pic_vector(img_name):
        img = Image.open(path+img_name).convert('RGB')
        image_tensor = preprocess(img)
        output_tensor = resnet(image_tensor.unsqueeze(0).to(device)).flatten().cpu()
        output_tensor.size()
        return output_tensor.numpy()    
    
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    resnet = resnet.to(device)    

    df_train_vectors = df_train['image'].progress_apply(pic_vector)

</div>

Таким образом мы получили массив векторов в порядке, в котором изображения находятся в train_data и test_images

## 4. Векторизация текстов

Следующий этап — векторизация текстов. Вы можете поэкспериментировать с несколькими способами векторизации текстов:

- tf-idf
- word2vec
- \*трансформеры (например Bert)

\* — если вы изучали трансформеры в спринте Машинное обучение для текстов.


Воспользуемся tf-idf

In [49]:
train_corpus = train_df['query_text'].values

In [50]:
nltk.download('stopwords')
#stop_words = set(stopwords.words('english'))

count_tf_idf = TfidfVectorizer(stop_words=stopwords.words('english'))
#(stop_words=set(stopwords.words('english')))
#s(stop_words=stopwords.words('english'))

text_vect = count_tf_idf.fit_transform(train_corpus)

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\Serge\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [51]:
text_vect.shape

(4002, 1155)

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
 👍 Вектора для текста получены с помощью TF-IDF. Здесь конечно есть очень тонкий момент, и в другом проекте это мог бы быть красный комментарий. Ты же помнишь, что векторайзеры мы обучаем на обучающем датасете и применяем к валидационому и тестовому. А сейчас ты сделал вектора с использованием всех частей датасета.  Но для этого проекта это должно быть допустимо, т.к. .утечка,повышающая оценку на валидации врядли что-то сильно  испортит.</div>

### 5. Объединение векторов
Подготовьте данные для обучения: объедините векторы изображений и векторы текстов с целевой переменной.

Чтобы объединить вектора, переведем их в один формат (np.array) и воспользуемся функцией np.hstack

In [52]:
image_vector = outputs_tensor.numpy()


In [53]:
text_vector = text_vect.toarray()

In [54]:
combined_array = np.concatenate((image_vector, text_vector), axis=1)

In [55]:
combined_array.shape

(4002, 1667)

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
 Вектора, полученные из текста и изображения объеденены👍 </div>

Суммы размерности сходятся: 4002  и 512+1155=1667

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Здесь напрашивается небольшой вывод, хотя бы с указанием размерностей полученных признаков.  
</div>

## 6. Обучение модели предсказания соответствия

Для обучения разделите датасет на тренировочную и тестовую выборки. Простое случайное разбиение не подходит: нужно исключить попадание изображения и в обучающую, и в тестовую выборки.
Для того чтобы учесть изображения при разбиении, можно воспользоваться классом [GroupShuffleSplit](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GroupShuffleSplit.html) из библиотеки sklearn.model_selection.

Код ниже разбивает датасет на тренировочную и тестовую выборки в пропорции 7:3 так, что строки с одинаковым значением 'group_column' будут содержаться либо в тестовом, либо в тренировочном датасете.

```
from sklearn.model_selection import GroupShuffleSplit
gss = GroupShuffleSplit(n_splits=1, train_size=.7, random_state=42)
train_indices, test_indices = next(gss.split(X=df.drop(columns=['target']), y=df['target'], groups=df['group_column']))
train_df, test_df = df.loc[train_indices], df.loc[test_indices]

```

Какую модель использовать — выберите самостоятельно. Также вам предстоит выбрать метрику качества либо реализовать свою.

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
В этом задании требуется выбрать основную метрику, которую ты будешь оптимизировать. Пожаулуйста выбери метрику и обоснуй свой выбор.  </div>

<div class="alert" style="background-color:#ead7f7;color:#8737bf">
    <font size="3"><b>Комментарий студента</b></font>
   
Будем оптимизировать RMSE см. описание ниже

</div>

In [56]:
target = train_df['avg_note']

In [57]:
gss = GroupShuffleSplit(n_splits=1, train_size=.7, random_state=6767)
train_indices, test_indices = next(gss.split(X=combined_array, y=target, groups=train_df['image']))
features_train, features_test = combined_array[train_indices], combined_array[test_indices]
target_train, target_test = target[train_indices], target[test_indices]

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Отлично, благодаря использованию GroupShuffleSplit ты исключаешь попадание одного изображения в обучающий и валидационный наборы.</div>

In [ ]:
scaler = StandardScaler()
scaler.fit(features_train)
features_train = scaler.transform(features_train)
features_test = scaler.transform(features_test)

Будем пробывать обучить линейную регрессию.

В качестве метрики выберем RMSE, поскольку эта метрика имеет ту же размерность, что и зависимая переменная, что облегчает интерпретацию и сопоставление с единицами измерения переменной.

Нам нужно стремиться к небольшому значению RMSE, чтобы быть уверенными в корректной работе модели.

In [ ]:
lr = LinearRegression()
lr.fit(features_train, target_train)

In [ ]:
y_pred = lr.predict(features_test)

rmse = mean_squared_error(target_test, y_pred) ** 0.5
rmse

In [ ]:
lin_reg = ElasticNet(alpha=0.001, random_state=45)

In [ ]:
lin_reg.fit(features_train, target_train)

In [ ]:
y_pred_reg = lin_reg.predict(features_test)

rmse_reg = mean_squared_error(target_test, y_pred_reg) ** 0.5
rmse_reg

### Выводы________________

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
Нельзя применять скаллер к предсказаниям, тем более обучать этот скаллер на этих же предсказаниях.

Да, линейная регрессия сильно переобучается. Но давай честно посмотрим правде в глаза и поймем насколько она плоха здесь.
</div>

<div class="alert" style="background-color:#ead7f7;color:#8737bf">
    <font size="3"><b>Комментарий студента</b></font>
   
Исправлено

</div>

Теперь создадим полносвязную нейронную сеть.

In [ ]:
X_train = torch.FloatTensor(np.array(X_train))
X_test = torch.FloatTensor(np.array(X_test))
y_train = torch.FloatTensor(np.array(y_train))
y_test = torch.FloatTensor(np.array(y_test))

In [ ]:
X_train.shape

In [ ]:
class Net(nn.Module):
    def __init__(self, n_in_neurons, n_hidden_neurons_1, n_hidden_neurons_2, n_hidden_neurons_3, n_hidden_neurons_4, n_out_neurons):
            super(Net, self).__init__()

            self.fc1 = nn.Linear(n_in_neurons, n_hidden_neurons_1)
            self.act1 = nn.ReLU()
            #self.do1 = nn.Dropout(p=0.7)
            self.fc2 = nn.Linear(n_hidden_neurons_1, n_hidden_neurons_2)
            self.act2 = nn.ReLU()
            self.fc3 = nn.Linear(n_hidden_neurons_2, n_hidden_neurons_3)
            self.act3 = nn.ReLU()
            self.fc4 = nn.Linear(n_hidden_neurons_3, n_hidden_neurons_4)
            self.act4 = nn.ReLU()
            self.fc5 = nn.Linear(n_hidden_neurons_4, n_out_neurons)
            self.act5 = nn.Sigmoid()

            nn.init.kaiming_uniform_(self.fc1.weight, mode='fan_in', nonlinearity='relu')
            nn.init.kaiming_uniform_(self.fc2.weight, mode='fan_in', nonlinearity='relu')
            nn.init.kaiming_uniform_(self.fc3.weight, mode='fan_in', nonlinearity='relu')
            nn.init.kaiming_uniform_(self.fc4.weight, mode='fan_in', nonlinearity='relu')
            nn.init.kaiming_uniform_(self.fc5.weight, mode='fan_in', nonlinearity='sigmoid')

    def forward(self, x):
        x = self.fc1(x)
        x = self.act1(x)
        #x = self.do1(x)
        x = self.fc2(x)
        x = self.act2(x)
        x = self.fc3(x)
        x = self.act3(x)
        x = self.fc4(x)
        x = self.act4(x)
        x = self.fc5(x)
        x = self.act5(x)

        return x

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Технически все верно. Но учитывая что даже простая модель переобучается я бы озаботился борьбой с переобучением. Возможно начал с более простых архитектур (один - два внутренних слоя, возможно поменьше нейронов).

Что касается регуляризации - стоит попробовать dropout (посмотри p в диапазоне 0,1-0,3), но например добавь слои прореживания после каждого внутреннего слоя.
    
Как альтернативу можно попробовать l1/l2 регуляризацию на веса, в pytorch она реализуется через параметр weight_decay оптимайзера.    

</div>

In [ ]:
n_in_neurons = 2800
n_hidden_neurons_1 = 1200
n_hidden_neurons_2 = 600
n_hidden_neurons_3 = 200
n_hidden_neurons_4 = 10
n_out_neurons = 1



net = Net(n_in_neurons, n_hidden_neurons_1, n_hidden_neurons_2, n_hidden_neurons_3, n_hidden_neurons_4, n_out_neurons)

#net = nn.Sequential(
#    nn.Linear(n_in_neurons, n_hidden_neurons_1),
#    nn.ReLU(),
#    nn.Linear(n_hidden_neurons_1, n_hidden_neurons_2),
#    nn.ReLU(),
#    nn.Linear(n_hidden_neurons_2, n_out_neurons)
#)

optimizer2 = torch.optim.SGD(net.parameters(), lr=1e-3, momentum=0.8)

optimizers = [optimizer2]

loss = nn.MSELoss()

In [ ]:
num_epochs = 900
batch_size = 150
num_batches = ceil(len(X_train)/batch_size)

for optimizer in optimizers:
    optimizer_rmse = []
    print(f'Оптимизатор {optimizer}:')

    for epoch in range(num_epochs):
        net.train()
        order = np.random.permutation(len(X_train))

        for batch_idx in range(num_batches):

            start_index = batch_idx * batch_size
            optimizer.zero_grad()

            batch_indexes = order[start_index : start_index + batch_size]
            X_batch = X_train[batch_indexes]
            y_batch = y_train[batch_indexes]

            preds = net.forward(X_batch.float()).flatten()

            loss_value = loss(preds, y_batch)

            loss_value.backward()

            optimizer.step()

        if epoch % 10 == 0 or epoch == num_epochs - 1:
            with torch.no_grad():
                net.eval()
                test_preds = net.forward(X_test).flatten()
                #print(test_preds.numpy())
                ans = round(float(torch.sqrt(loss(test_preds, y_test))), 2)
                print(f'Метрика RMSE на {epoch} итерации =', ans)
                optimizer_rmse.append(ans)
    print()
    print(f'Лучшая метрика RMSE = {min(optimizer_rmse)} обнаружена на {optimizer_rmse.index(min(optimizer_rmse))} эпохе')
    print(f'Средняя метрика RMSE = {round(np.mean(optimizer_rmse), 2)}')
    print('---------------------------------------------------------------------------------------------', '\n')

## 7. Тестирование модели

Настало время протестировать модель. Для этого получите эмбеддинги для всех тестовых изображений из папки `test_images`, выберите случайные 10 запросов из файла `test_queries.csv` и для каждого запроса выведите наиболее релевантное изображение. Сравните визуально качество поиска.

In [ ]:
#создаем вектор изображений
image_vectors = []
test_queries['image'].progress_apply(lambda x: image_vect(x, PATH_TEST))
outputs_tensor_test = torch.stack(image_vectors)

In [ ]:
outputs_tensor_test.shape

In [ ]:
df_test_images = pd.read_csv(os.path.join(PATH, 'test_images.csv'))

Пишем функцию для тестирования работы модели — выдачи изображения по запросу.

In [ ]:
def testing(text):
    print(f'Описание: {text}')
    for word in text.split():
        if word in stop_kids:
            return print('WARNING: This image is unavailable in your country in compliance with local laws.')
    Xids_test = count_tf_idf.transform([text]).toarray()
    best = 0
    best_index = 0
    pred_list = []

    for i in range(len(outputs_tensor_test)):

        features_test = pd.DataFrame(outputs_tensor_test[i]).T.merge(pd.DataFrame(Xids_test), left_index=True, \
                                                      right_index=True, suffixes=(None, '_t'))

        features_test.columns = features_test.columns.astype(str)
        target_pred_test = cb.predict(features_test.to_numpy())
        pred_list.append(target_pred_test[0])

        if target_pred_test > best:
            best = target_pred_test
            best_index = i

    print(f'Лучшая вероятность: {best}')
    print(f'Лучший индекс: {best_index}')
    print(f'Лучшее изображение:')
    img = mpimg.imread(os.path.join(PATH, 'test_images/') + test_queries.image[best_index])
    imgplot = plt.imshow(img)
    plt.show()

In [ ]:
queries_for_test = test_queries['query_text'].sample(10, random_state=123)
queries_for_test

In [ ]:
for query in queries_for_test:
    testing(query)

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
Небольшая формальность - пожалуйста добавь в код функции инфереса, чтобы она кроме изображения и текста  выводила найденную меру близости.

Ну и в соответствии с требованиями пожалуйста покажи найденные фото.
</div>

<div class="alert" style="background-color:#ead7f7;color:#8737bf">
    <font size="3"><b>Комментарий студента</b></font>
   
Добавлено

</div>

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Функции написаны и со своей задачей справляются. Принимают на вход текст и находят изображение. Отрабатывают контроль "запрещенного контента".

Да, мы столкнулись с тем, что модель выдает одно и то же изображение. Это связано с тем, что модель похоже "застряла" в локальном минимуме. Одна из причин такого поведения - использование векторов полученных с помощью TF-IDF, в которых очень много нулей и редко встречаются числа, отличные от нуля. Поэтому модели крайне сложно найти взаимосвязь между такими разноплановыми признаками и таргетом, она слишком быстро переобучается на какие-то другие признаки (тем более и данных у нас немного).
    
Здесь совет только один - поработать с большим объемом датасета + больше экспериментов с архитетурой модели и попытками обучить её. Может получиться (а может и нет).     
</div>

<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Идеи и альтернативные подходы</b>
  Если хочешь получить хорошо работающее решение без обучения, то можешь провести эксперимент с нейронной сетью CLIP от OPEN AI (это не обязательно, только если хочешь добиться хороших результатов в этом проекте).

  Про эту нейронку много где написано, но можешь посмотреть например вот здесь  https://habr.com/ru/post/539312/.

  А для написания  реализации наиболее удобной лично для меня оказалась библиотека sbert. https://www.sbert.net/examples/applications/image-search/README.html  
    
    
<hr>    
    
Также неплохо работает следующий подход без предобученных моделей (к тому же легко реализуемый).
    
    
Вместо того, чтобы подавать на вход модели изображение + текст и получать меру соответствия, поступаем по другому. 
    
На вход модели подаем вектор описания, а на выходе учимся предсказывать вектор изображения. 
    
По поводу метрик тут ситуация следующая. Точно работает следующий подход
    
  - берем метрику mse, но тогда нужно чтобы модель её уменьшала. А значит в обучении используем только пары заведомо сответствующие друг другу

Можно попробовать использовать косинусное расстояние, она позволяет использовать все данные (если есть соответствие то выдает 1, если нет то 0). 
    
Также можно попробовать усложнить модель, например и изображения и тексты отправлять в пространство размерностью 256 и уже в этом пространстве учиться их "сближать".
    
По поводу методов векторизации: я бы использовать resnet18 для изображений (размерность эмбеддинга избражения 512) и БЕРТ для текстов (768). 
    
***
    
А еще я видел удачный проект, где для векторизации изображений использовалось предсказание Resnet в котором не отключался последний слой. Получился вектор предсказаний верояятности что изображение принадлежит к каждому из 1000 классов Imagenet. И в сочетании с подходом "мешок слов" для текстов это может дать неплохой результа. Такая модель находит людей, собак, машины, снег и т.д.  
   
</div>

## 8. Выводы

- Jupyter Notebook открыт
- Весь код выполняется без ошибок
- Ячейки с кодом расположены в порядке исполнения
- Исследовательский анализ данных выполнен
- Проверены экспертные оценки и краудсорсинговые оценки
- Из датасета исключены те объекты, которые выходят за рамки юридических ограничений
- Изображения векторизованы
- Текстовые запросы векторизованы
- Данные корректно разбиты на тренировочную и тестовую выборки
- Предложена метрика качества работы модели
- Предложена модель схожести изображений и текстового запроса
- Модель обучена
- По итогам обучения модели сделаны выводы
- Проведено тестирование работы модели
- По итогам тестирования визуально сравнили качество поиска

<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Заключительный комментарий</b>

Подведем итоги. Я вижу, что в целом с проектом ты справляешься.
    
Проведено исследование и предобработка данных. Выполнена векторизация изображений и текстовых описаний. Обучены модели для предсказания степени соответствия, а также написана функция для инференса и с её помощью проведено тестирование.
 
    
Что нужно доделать обязательно:
    
 - пожалуйста добавь выводы после знакомства с данными
 - пожалуйста используй альтернативный подход к объединению экспертных и краудсорс оценок (не соединение через inner)
 - нужно обосновать выбор оптимизируемой метрики
 - нельзя применять скаллер к предсказаниям, тем более обучать его на них
 - пожалуйста доработай функцию для инференса
    
Желтые комментарии на твое усмотрение. Но учитывая что наш подход "не взлетел" и нужно хоть что-то показать заказчику, попробуй показать возможный инференс с помщью CLIP. С библиотекой sentence-transformers это делается буквально в 10 строк.    
Жду твою работу на повторное ревью.
</div>